[quant][graphmode][fx] Add support for packed params in state_dict

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52802 [do not land] ad hoc test for conv serialization
* **#51639 [quant][graphmode][fx] Add support for packed params in state_dict**
* #52787 [quant] Fix conv packed param serialization in state_dict

Summary:
Previously we did not store packed params such as LinearPackedParams, Conv2dPackedParams in state_dict, therefore we can't really serialize/derserilize a quantized GraphModule with `torch.save(m.state_dict())` and `torch.load(state_dict)`, which is the recommended ser/deser for normal nn.Modules. This PR added packed param to state_dict by overwriting `save_to_state_dict` and `load_from_state_dict` methods.

An alternative we considered is too unpack packed params and save the quantized weight/bias instead, however there will be some difficulty when loading the weights since we need to manually match weight with the corresponding bias and other parameters before we can call packing ops like linear_prepack and conv2d_prepack.

In a follow up PR we will change the conv and linear module to save packed params directly as well. ETA for the PR is within 1 week after this is landed.


Test Plan:
python test/test_quantization.py TestQuantizeFx.test_state_dict
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26228185](https://our.internmc.facebook.com/intern/diff/D26228185)

jerryzh168 committed:
[quant][graphmode][fx] Add support for packed params in state_dict

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for packed params in state_dict"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for packed params in state_dict"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26228185](https://our.internmc.facebook.com/intern/diff/D26228185)

[ghstack-poisoned]

supriyar commented on test/quantization/test_quantize_fx.py:
```diff
@@ -1592,6 +1592,57 @@ def forward(self, x):
+        m = M2().eval()
+        m = prepare_fx(m, qconfig_dict)
+        m = convert_fx(m)
+        m.load_state_dict(state_dict)
```
Can we run forward on the 2nd model with same set of inputs to compare with 1st model?
This will ensure that we loaded all quantization related params correctly from the state_dict

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for packed params in state_dict"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26228185](https://our.internmc.facebook.com/intern/diff/D26228185)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for packed params in state_dict"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26228185](https://our.internmc.facebook.com/intern/diff/D26228185)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for packed params in state_dict"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26228185](https://our.internmc.facebook.com/intern/diff/D26228185)

[ghstack-poisoned]

vkuzo commented on test/quantization/test_quantize_fx.py:
```diff
@@ -1592,6 +1594,68 @@ def forward(self, x):
         assert hasattr(m, "mods1_1_packed_weight_0")
         assert hasattr(m, "mods2_packed_weight_0")
 
+    def test_state_dict(self):
```
would it also make sense to test saving to disk and successfully loading back?

jerryzh168 replied:
Oh sure, yeah I think that would be useful

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -903,7 +904,7 @@ def load_arg_simple(a: Argument) -> Argument:
         # removes qconfig and activation_post_process modules
         _remove_qconfig(model)
-        model = GraphModule(model, act_post_process_removed_graph)
+        model = QuantizedGraphModule(model, act_post_process_removed_graph)
```
could we update the type hint on _convert with the new type?

vkuzo commented:
overall I think if we align that we want to do this, this would be a good first step.  Ideally we would get some feedback from PT core if they're ok with this.  If everyone is good, it would be good if we also adjust the quantized conv and linear behavior in the same stack.

vkuzo commented:
could we also add a summary and test plan, just so people who find this PR later have a sense of what's going on?

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for packed params in state_dict"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26228185](https://our.internmc.facebook.com/intern/diff/D26228185)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for packed params in state_dict"


Summary:
Previously we did not store packed params such as LinearPackedParams, Conv2dPackedParams in state_dict, therefore we can't really serialize/derserilize a quantized GraphModule with `torch.save(m.state_dict())` and `torch.load(state_dict)`, which is the recommended ser/deser for normal nn.Modules. This PR added packed param to state_dict by overwriting `save_to_state_dict` and `load_from_state_dict` methods.

An alternative we considered is too unpack packed params and save the quantized weight/bias instead, however there will be some difficulty when loading the weights since we need to manually match weight with the corresponding bias and other parameters before we can call packing ops like linear_prepack and conv2d_prepack.

In a follow up PR we will change the conv and linear module to save packed params directly as well


Test Plan:
python test/test_quantization.py TestQuantizeFx.test_state_dict
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26228185](https://our.internmc.facebook.com/intern/diff/D26228185)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for packed params in state_dict"


Summary:
Previously we did not store packed params such as LinearPackedParams, Conv2dPackedParams in state_dict, therefore we can't really serialize/derserilize a quantized GraphModule with `torch.save(m.state_dict())` and `torch.load(state_dict)`, which is the recommended ser/deser for normal nn.Modules. This PR added packed param to state_dict by overwriting `save_to_state_dict` and `load_from_state_dict` methods.

An alternative we considered is too unpack packed params and save the quantized weight/bias instead, however there will be some difficulty when loading the weights since we need to manually match weight with the corresponding bias and other parameters before we can call packing ops like linear_prepack and conv2d_prepack.

In a follow up PR we will change the conv and linear module to save packed params directly as well


Test Plan:
python test/test_quantization.py TestQuantizeFx.test_state_dict
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26228185](https://our.internmc.facebook.com/intern/diff/D26228185)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for packed params in state_dict"


Summary:
Previously we did not store packed params such as LinearPackedParams, Conv2dPackedParams in state_dict, therefore we can't really serialize/derserilize a quantized GraphModule with `torch.save(m.state_dict())` and `torch.load(state_dict)`, which is the recommended ser/deser for normal nn.Modules. This PR added packed param to state_dict by overwriting `save_to_state_dict` and `load_from_state_dict` methods.

An alternative we considered is too unpack packed params and save the quantized weight/bias instead, however there will be some difficulty when loading the weights since we need to manually match weight with the corresponding bias and other parameters before we can call packing ops like linear_prepack and conv2d_prepack.

In a follow up PR we will change the conv and linear module to save packed params directly as well


Test Plan:
python test/test_quantization.py TestQuantizeFx.test_state_dict
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26228185](https://our.internmc.facebook.com/intern/diff/D26228185)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for packed params in state_dict"


Summary:
Previously we did not store packed params such as LinearPackedParams, Conv2dPackedParams in state_dict, therefore we can't really serialize/derserilize a quantized GraphModule with `torch.save(m.state_dict())` and `torch.load(state_dict)`, which is the recommended ser/deser for normal nn.Modules. This PR added packed param to state_dict by overwriting `save_to_state_dict` and `load_from_state_dict` methods.

An alternative we considered is too unpack packed params and save the quantized weight/bias instead, however there will be some difficulty when loading the weights since we need to manually match weight with the corresponding bias and other parameters before we can call packing ops like linear_prepack and conv2d_prepack.

In a follow up PR we will change the conv and linear module to save packed params directly as well


Test Plan:
python test/test_quantization.py TestQuantizeFx.test_state_dict
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26228185](https://our.internmc.facebook.com/intern/diff/D26228185)

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -1546,15 +1548,17 @@ def forward(self, x):
+        scripted.mods1_0_packed_weight_0 = m.state_dict()["mods1_0_packed_weight_0"]
+        non_packed_weight_keys = [key for key in keys if "_packed_weight" not in key]
+        self.assertTrue(
+            set(scripted_keys) == set(non_packed_weight_keys),
```
I thought the scripted model also has the packed weight attr. Is my understanding incorrect?

raghuramank100 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -1546,15 +1548,17 @@ def forward(self, x):
-        assert hasattr(m, "mods1_1_zero_point_0")
-        assert hasattr(m, "mods2_scale_0")
-        assert hasattr(m, "mods2_zero_point_0")
+        scripted.mods1_0_packed_weight_0 = m.state_dict()["mods1_0_packed_weight_0"]
```
Why does the scripted model not have packed weight in its state dict? What is the work needed to enable that?

jerryzh168 replied:
I heard that state_dict is not a supported api in scripted model. cc @jamesr66a @suo to confirm

vkuzo commented:
> In a follow up PR we will change the conv and linear module to save packed params directly as well

What's the ETA for the follow-up PRs to update conv and linear modules to conform, and can we document that on the diff summary?  Usually it's a good practice to either land things like this in a single PR (best), or at least a stack of PRs all landing at the same time.  An exception could be made in this case since no one is using this path yet, but would be great to document the plan.

albanD commented on torch/quantization/fx/graph_module.py:
```diff
@@ -47,8 +44,34 @@ def __deepcopy__(self, memo):
+            if "_packed_weight" in attr_name and \
+               isinstance(getattr(self, attr_name), torch._C.ScriptObject):  # type: ignore
+                packed_weight = getattr(self, attr_name)
+                destination[prefix + attr_name] = packed_weight
```
You are missing the handling of the `keep_vars` param here I think. If this is a Tensor, it should be `packed_weight if keep_vars else packed_weight.detach()`.

jerryzh168 replied:
`packed_weight` is not a Tensor, it's a TorchBind object, do we still need to call detach() on it?

albanD replied:
Do they contain Tensors that can require gradients? If so, you should detach all of them yes.

jerryzh168 replied:
`packed_weight` has type ConvPackedParams/LinearPackedParams, and they do not contain any tensors, but they are generated from Tensors (by packing them).

albanD commented on torch/quantization/fx/graph_module.py:
```diff
@@ -47,8 +44,34 @@ def __deepcopy__(self, memo):
+        attrs_to_pop = []
+        for attr_name in state_dict:
+            if "_packed_weight" in attr_name and isinstance(state_dict[attr_name], torch._C.ScriptObject):  # type: ignore
+                setattr(self, attr_name, state_dict[attr_name])
```
Should these be registered as Parameters in any way?

jerryzh168 replied:
these are not trainable. packed params are the packed quantized weight + bias after quantization. for example, it's generated by:
```
self._packed_weight = torch.ops.quantized.linear_prepack(quantized_weight, bias)
```

ideally they should be registered as buffers, but currently buffer only support Tensor and Module I think.

albanD replied:
ok!

albanD commented on torch/quantization/fx/graph_module.py:
```diff
@@ -47,8 +44,34 @@ def __deepcopy__(self, memo):
+    def __init__(self, root: Union[torch.nn.Module, Dict[str, Any]], graph: Graph):
+        super().__init__(root, graph)
+
+    def _save_to_state_dict(self, destination, prefix, keep_vars):
```
Why not use the hooks instead of overriding the function itself: https://github.com/pytorch/pytorch/blob/8af648354f4ae8e00576030f558ee08deb1c4acd/torch/nn/modules/module.py#L1008 ?

jerryzh168 replied:
sure, I can use hooks

jerryzh168 replied:
why there is no self in the prehook? https://github.com/pytorch/pytorch/blob/8af648354f4ae8e00576030f558ee08deb1c4acd/torch/nn/modules/module.py#L1127
I will need to modify self in this case.

Can I use hook for _save_to_state_dict but overload _load_from_state_dict?

albanD replied:
>  I will need to modify self in this case.

Why? The current code doesn't right?
Also modifying the module itself when it is saved sounds wrong no?

jerryzh168 replied:
I mean I need to modify the module when loading from state_dict

albanD replied:
Ho that's interesting that this is not there... I guess you're only supposed to change the state dict itself from there?
cc @jbschlosser

albanD commented on test/quantization/test_quantize_fx.py:
```diff
@@ -1624,6 +1628,94 @@ def forward(self, x):
+        checkModel(m, data, ref_weight, ref_bias, ref_res)
+
+        # check quantized linear works
+        # m = M1().eval()
```
Should this be removed?

albanD commented on test/quantization/test_quantize_fx.py:
```diff
@@ -1624,6 +1628,94 @@ def forward(self, x):
+        m = prepare_fx(m, qconfig_dict)
+        m = convert_fx(m)
+        with TemporaryFileName() as fname:
+            print(state_dict)
```
nit: to remove

albanD commented on torch/quantization/fx/graph_module.py:
```diff
@@ -47,8 +44,34 @@ def __deepcopy__(self, memo):
+    def _save_to_state_dict(self, destination, prefix, keep_vars):
+        super()._save_to_state_dict(destination, prefix, keep_vars)
+        for attr_name in dir(self):
+            if "_packed_weight" in attr_name and \
```
I am not familiar with the quantization process but is "weight" always used here? Or can it be anything else?

jerryzh168 replied:
do you mean the "_packed_weight" name?

albanD replied:
Yes.
In particular, the `_packed` part sounds like it can be common, but I wasn't sure about the weight?
Also you might want to match the beginning of the string here as well to avoid matching things like "my_custom_packed_weight".

jerryzh168 replied:
sure, I have the same concerns, I can change that to a startswith check

albanD reviewed (commented):
It looks ok. I think that using the hook version for the state dict changes would be cleaner if possible.

jerryzh168 commented:
> > In a follow up PR we will change the conv and linear module to save packed params directly as well
> 
> What's the ETA for the follow-up PRs to update conv and linear modules to conform, and can we document that on the diff summary? Usually it's a good practice to either land things like this in a single PR (best), or at least a stack of PRs all landing at the same time. An exception could be made in this case since no one is using this path yet, but would be great to document the plan.

Sure. ETA will be within 1 week after this lands, updated the ETA in description.

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for packed params in state_dict"


Summary:
Previously we did not store packed params such as LinearPackedParams, Conv2dPackedParams in state_dict, therefore we can't really serialize/derserilize a quantized GraphModule with `torch.save(m.state_dict())` and `torch.load(state_dict)`, which is the recommended ser/deser for normal nn.Modules. This PR added packed param to state_dict by overwriting `save_to_state_dict` and `load_from_state_dict` methods.

An alternative we considered is too unpack packed params and save the quantized weight/bias instead, however there will be some difficulty when loading the weights since we need to manually match weight with the corresponding bias and other parameters before we can call packing ops like linear_prepack and conv2d_prepack.

In a follow up PR we will change the conv and linear module to save packed params directly as well. ETA for the PR is within 1 week after this is landed.


Test Plan:
python test/test_quantization.py TestQuantizeFx.test_state_dict
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26228185](https://our.internmc.facebook.com/intern/diff/D26228185)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for packed params in state_dict"


Summary:
Previously we did not store packed params such as LinearPackedParams, Conv2dPackedParams in state_dict, therefore we can't really serialize/derserilize a quantized GraphModule with `torch.save(m.state_dict())` and `torch.load(state_dict)`, which is the recommended ser/deser for normal nn.Modules. This PR added packed param to state_dict by overwriting `save_to_state_dict` and `load_from_state_dict` methods.

An alternative we considered is too unpack packed params and save the quantized weight/bias instead, however there will be some difficulty when loading the weights since we need to manually match weight with the corresponding bias and other parameters before we can call packing ops like linear_prepack and conv2d_prepack.

In a follow up PR we will change the conv and linear module to save packed params directly as well. ETA for the PR is within 1 week after this is landed.


Test Plan:
python test/test_quantization.py TestQuantizeFx.test_state_dict
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26228185](https://our.internmc.facebook.com/intern/diff/D26228185)

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/graph_module.py:
```diff
@@ -47,8 +44,33 @@ def __deepcopy__(self, memo):
+            packed_weight = getattr(self, attr_name)
+            destination[prefix + attr_name] = packed_weight
+
+class QuantizedGraphModule(GraphModule):
```
would it be possible to add a short docblock on why this class existis, so a code reader could understand it without opening this PR to get the context?

jerryzh168 replied:
Sure

vkuzo commented on torch/quantization/fx/graph_module.py:
```diff
@@ -47,8 +44,33 @@ def __deepcopy__(self, memo):
+class QuantizedGraphModule(GraphModule):
+    def __init__(self, root: Union[torch.nn.Module, Dict[str, Any]], graph: Graph):
+        super().__init__(root, graph)
+        self._register_state_dict_hook(_save_packed_weight)
```
just checking, would this work if the module is scripted?

jerryzh168 replied:
I don't think hooks are supported for script modules. cc @albanD to confirm

albanD replied:
I don't know about state dict ones. But regular hooks are not supported, so I guess not.

vkuzo replied:
ah, I see.  Would it make sense to add some test cases for making sure the state dict loading and saving works after scripting, just to be safe?

jerryzh168 replied:
raghu had a similar comment above, but I remember from james/michael that state_dict is not a supported api in script modules

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for packed params in state_dict"


Summary:
Previously we did not store packed params such as LinearPackedParams, Conv2dPackedParams in state_dict, therefore we can't really serialize/derserilize a quantized GraphModule with `torch.save(m.state_dict())` and `torch.load(state_dict)`, which is the recommended ser/deser for normal nn.Modules. This PR added packed param to state_dict by overwriting `save_to_state_dict` and `load_from_state_dict` methods.

An alternative we considered is too unpack packed params and save the quantized weight/bias instead, however there will be some difficulty when loading the weights since we need to manually match weight with the corresponding bias and other parameters before we can call packing ops like linear_prepack and conv2d_prepack.

In a follow up PR we will change the conv and linear module to save packed params directly as well. ETA for the PR is within 1 week after this is landed.


Test Plan:
python test/test_quantization.py TestQuantizeFx.test_state_dict
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26228185](https://our.internmc.facebook.com/intern/diff/D26228185)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add support for packed params in state_dict"


Summary:
Previously we did not store packed params such as LinearPackedParams, Conv2dPackedParams in state_dict, therefore we can't really serialize/derserilize a quantized GraphModule with `torch.save(m.state_dict())` and `torch.load(state_dict)`, which is the recommended ser/deser for normal nn.Modules. This PR added packed param to state_dict by overwriting `save_to_state_dict` and `load_from_state_dict` methods.

An alternative we considered is too unpack packed params and save the quantized weight/bias instead, however there will be some difficulty when loading the weights since we need to manually match weight with the corresponding bias and other parameters before we can call packing ops like linear_prepack and conv2d_prepack.

In a follow up PR we will change the conv and linear module to save packed params directly as well. ETA for the PR is within 1 week after this is landed.


Test Plan:
python test/test_quantization.py TestQuantizeFx.test_state_dict
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26228185](https://our.internmc.facebook.com/intern/diff/D26228185)

[ghstack-poisoned]
